bfvmm/src/hve/arch/intel_x64/xen/xen_op.cpp

```diff
@@ -357,6 +349,208 @@ xen_op_handler::exit_handler(
     return false;
 }
 
+// -----------------------------------------------------------------------------
+// xAPIC
+// -----------------------------------------------------------------------------
+
+//bool
+//xen_op_handler::xapic_handle_read(
+//    gsl::not_null<vcpu_t *> vcpu,
+//    eapis::intel_x64::ept_violation_handler::info_t &info)
+//{
+////    auto hkv = vcpu_cast(vcpu);
+////    if (bfn::upper(info.gpa) != hkv->lapic_base()) {
+////        return false;
+////    }
+////
+////    auto idx = bfn::lower(info.gpa) >> 2;
+////    bfdebug_nhex(0, "xapic read", idx << 2);
+////    bfdebug_subnhex(0, "val", m_vcpu->lapic_read(idx));
+////
+////    const auto len = ::intel_x64::vmcs::vm_exit_instruction_length::get();
+////    const auto rip = ::intel_x64::vmcs::guest_rip::get();
+////
+////    auto itr = m_xapic_rip_cache.find(rip);
+////    if (itr == m_xapic_rip_cache.end()) {
+////        auto ump = hkv->map_gva_4k<uint8_t>(rip, len);
+////        if (!ump) {
+////            throw std::runtime_error("handle_xapic_write::map_gva_4k failed");
+////        }
+////
+////        m_xapic_rip_cache[rip] = std::move(ump);
+////        itr = m_xapic_rip_cache.find(rip);
+////    }
+////
+////    const auto buf = itr->second.get();
+////    printf("received xapic insn: ");
+////    for (auto i = 0; i < len; i++) {
+////        printf("%02x", buf[i]);
+////    }
+////    printf("\n");
+//
+//    return false;
+//}
+
+uint64_t
+src_op_value(gsl::not_null<vcpu_t *> vcpu, int64_t src_op)
+{
+    switch (src_op) {
+        case hyperkernel::intel_x64::insn_decoder::eax:
+            return vcpu->rax() & 0xFFFFFFFFU;
+        case hyperkernel::intel_x64::insn_decoder::ecx:
+            return vcpu->rcx() & 0xFFFFFFFFU;
+        case hyperkernel::intel_x64::insn_decoder::edx:
+            return vcpu->rdx() & 0xFFFFFFFFU;
+        case hyperkernel::intel_x64::insn_decoder::ebx:
+            return vcpu->rbx() & 0xFFFFFFFFU;
+        case hyperkernel::intel_x64::insn_decoder::esp:
+            return vcpu->rsp() & 0xFFFFFFFFU;
+        case hyperkernel::intel_x64::insn_decoder::ebp:
+            return vcpu->rbp() & 0xFFFFFFFFU;
+        case hyperkernel::intel_x64::insn_decoder::esi:
+            return vcpu->rsi() & 0xFFFFFFFFU;
+        case hyperkernel::intel_x64::insn_decoder::edi:
+            return vcpu->rdi() & 0xFFFFFFFFU;
+    }
+
+    throw std::invalid_argument("invalid reg");
+}
+
+static void print_insn(const unsigned char *buf, size_t len)
+{
+    for (auto i = 0; i < len; i++) {
+        printf("%02x", buf[i]);
+    }
+}
+
+void
+xen_op_handler::xapic_handle_write_icr(uint64_t low)
```

Comment:
I am assuming these can all be removed? Also, all of the APIC logic should be in the APIC class. I see the vAPIC a Hyperkernel thing, not a PV interface thing. So the RDMSR and WRMSR functions should all be in the APIC class and self contained.

Reply:
I don't agree. I think the object that represents a local APIC should have absolutely no notion of an exit or vtx or anything like that. I'm fine with making an apic handler class though

Reply:
Yeah, I don't really care about the details for now of how the lapic is represented internally. If you want to have another class that abstracts APIC stuff that's fine, but what I really care about is the lapic.cpp in the hyperkernel is (like the rest of the EAPIs code) is an lapic_handler that exposes an API that everything else uses. I don't want APIC logic (including exits) sprinkled throughout the code.